src/turbine/turbine_tree.zig

```diff
@@ -0,0 +1,804 @@
+    reference_count: AtomicUsize,
+
+    /// The maximum number of nodes each node should retransmit to
+    pub const DATA_PLANE_FANOUT: usize = 200;
+    /// The maximum depth of the TurbineTree (0->1->2->3)
+    /// Fanout of 200 and max depth of 4 allows for ~200^3 (8 million) nodes
+    pub const MAX_TURBINE_TREE_DEPTH: usize = 4;
+    /// The maximum number of nodes per IP address
+    /// When this limit is reached, the nodes contact info is removed so that
+    /// the shuffle is deterministic but the node is not used for retransmit
+    pub const MAX_NODES_PER_IP_ADDRESS: usize = 10;
+
+    /// Nodes in the TurbineTree may be identified by solely their
+    /// pubkey if they are not in the gossip table or their contact info
+    /// is not known
+    pub const NodeId = union(enum) {
+        contact_info: ThreadSafeContactInfo,
+        pubkey: Pubkey,
+    };
+
+    /// A node in the TurbineTree
+    pub const Node = struct {
+        id: NodeId,
+        stake: u64,
+
+        pub fn pubkey(self: Node) Pubkey {
+            return switch (self.id) {
+                .contact_info => |ci| ci.pubkey,
+                .pubkey => |pk| pk,
+            };
+        }
+
+        pub fn contactInfo(self: Node) ?ThreadSafeContactInfo {
+            return switch (self.id) {
+                .contact_info => |ci| ci,
+                .pubkey => null,
+            };
+        }
+
+        pub fn tvuAddress(self: Node) ?SocketAddr {
+            return switch (self.id) {
+                .contact_info => |ci| ci.tvu_addr,
+                .pubkey => null,
+            };
+        }
+
+        pub fn fromContactInfo(ci: ThreadSafeContactInfo) Node {
+            return .{ .id = .contact_info(ci), .stake = ci.stake };
+        }
+    };
+
+    pub fn initForRetransmit(
+        allocator: std.mem.Allocator,
+        my_contact_info: ThreadSafeContactInfo,
+        gossip_table_rw: *RwMux(GossipTable),
+        staked_nodes: *const std.AutoArrayHashMapUnmanaged(Pubkey, u64),
+        use_stake_hack_for_testing: bool,
+    ) !TurbineTree {
+        const tvu_peers = try getTvuPeers(
+            allocator,
+            my_contact_info,
+            gossip_table_rw,
+        );
+        defer tvu_peers.deinit();
+
+        const nodes = try getNodes(
+            allocator,
+            my_contact_info,
+            tvu_peers.items,
+            staked_nodes,
+            use_stake_hack_for_testing,
+        );
+        errdefer nodes.deinit();
+
+        var index = std.AutoArrayHashMap(Pubkey, usize).init(allocator);
+        errdefer index.deinit();
+        for (nodes.items, 0..) |node, i| try index.put(node.pubkey(), i);
+
+        var node_stakes = try std.ArrayList(u64).initCapacity(allocator, nodes.items.len);
+        defer node_stakes.deinit();
+        for (nodes.items) |node| node_stakes.appendAssumeCapacity(node.stake);
+
+        const weighted_shuffle = try WeightedShuffle.init(allocator, node_stakes.items);
```

Comment:
```suggestion
        const node_stakes = try allocator.alloc(u64, nodes.items.len);
        defer allocator.free(node_stakes);
        for (node_stakes, nodes.items) |*stake, node| stake.* = node.stake;

        const weighted_shuffle = try WeightedShuffle.init(allocator, node_stakes);
```
Tell me if this isn't reasonable, but if we changed `nodes` to be a `std.MultiArrayList`, we could change this to simply be:
```suggestion
        const weighted_shuffle = try WeightedShuffle.init(allocator, nodes.items(.stake));
```
From a cursory look, it seems like it could also be taken advantage in other places.

Reply:
I've thought about this a little, I think it is not necessary here as there will be no impact on performance and I don't think it makes the code any clearer. We need to iterate over the nodes to create the index so using nodes.items(.stake) wont add a performance benefit over just moving `node_stakes.appendAssumeCapacity(node.stake);` into the loop which creates the node index. In terms of subsequent access when computing retransmit children, again I don't see a performance or syntactic benefit to using a multi array list. 7e9c4499